src/liballoc/collections/btree/navigate.rs

```diff
@@ -3,118 +3,63 @@ use core::ptr;
-    { unsafe fn $name:ident : $next_kv:ident $next_edge:ident $initial_leaf_edge:ident } => {
-        /// Given a leaf edge handle into an owned tree, returns a handle to the next
-        /// leaf edge and the key and value between these edges, while deallocating
-        /// any node left behind.
+macro_rules! def_next_kv_uncheched_dealloc {
+    { unsafe fn $name:ident : $adjacent_kv:ident } => {
+        /// Given a leaf edge handle into an owned tree, returns a handle to the next KV,
+        /// while deallocating any node left behind.
         /// Unsafe for two reasons:
-        /// - the caller must ensure that the given leaf edge has a successor;
-        /// - the node pointed at by the given handle, and its ancestors, may be deallocated,
+        /// - The caller must ensure that the leaf edge is not the last one in the tree.
+        /// - The node pointed at by the given handle, and its ancestors, may be deallocated,
         ///   while the reference to those nodes in the surviving ancestors is left dangling;
-        ///   thus using the returned handle is dangerous.
+        ///   thus using the returned handle to navigate further is dangerous.
         unsafe fn $name <K, V>(
             leaf_edge: Handle<NodeRef<marker::Owned, K, V, marker::Leaf>, marker::Edge>,
-        ) -> (Handle<NodeRef<marker::Owned, K, V, marker::Leaf>, marker::Edge>, K, V) {
-            let mut cur_handle = match leaf_edge.$next_kv() {
-                Ok(leaf_kv) => {
-                    let k = ptr::read(leaf_kv.reborrow().into_kv().0);
-                    let v = ptr::read(leaf_kv.reborrow().into_kv().1);
-                    let next_leaf_edge = leaf_kv.$next_edge();
-                    return (next_leaf_edge, k, v);
-                }
+        ) -> Handle<NodeRef<marker::Owned, K, V, marker::LeafOrInternal>, marker::KV> {
+            let mut parent_edge = match leaf_edge.$adjacent_kv() {
+                Ok(leaf_kv) => return leaf_kv.forget_node_type(),
                 Err(last_edge) => {
                     unwrap_unchecked(last_edge.into_node().deallocate_and_ascend())
                 }
             };
 
             loop {
-                cur_handle = match cur_handle.$next_kv() {
-                    Ok(internal_kv) => {
-                        let k = ptr::read(internal_kv.reborrow().into_kv().0);
-                        let v = ptr::read(internal_kv.reborrow().into_kv().1);
-                        let next_internal_edge = internal_kv.$next_edge();
-                        let next_leaf_edge = next_internal_edge.descend().$initial_leaf_edge();
-                        return (next_leaf_edge, k, v);
-                    }
+                parent_edge = match parent_edge.$adjacent_kv() {
+                    Ok(internal_kv) => return internal_kv.forget_node_type(),
                     Err(last_edge) => {
                         unwrap_unchecked(last_edge.into_node().deallocate_and_ascend())
                     }
```

Comment:
Same as before, please deduplicate here.

Reply:
This might be controversial. `deallocate_and_ascend` is defined separately for leaves and internal nodes because the block deallocated is actually different in type and size. So now a `force()` is needed in node.rs to choose between them. More code collapsed, but perhaps hairy. And then the last stage of deallocation in a map's Drop also collapses, for better or worse.

Reply:
Hm, so are you saying the types wouldn't match up? AFAICT, the code is identical when just looking at it -- I think in practice, the operation is the important bit so I would say that we should deduplicate as I've suggested.

Reply:
I don't understand what you mean with "types not matching up" and "identical" (internal nodes have an edge array, leaves don't, so they match but are not identical) nor with "operation", but you can see the proposed result right now.